File: agent/sysmon-agent.py
#!/usr/bin/env python3
"""sysmon-agent (Python) — CPU/RAM metrics endpoint for the Ulanzi System Monitor plugin.

Linux only (reads /proc). Stdlib only — no pip installs, so it runs on any host
with python3. Serves the SAME JSON contract as the Node agent, so the plugin
treats them identically. Use this on Linux hosts that don't have Node.

Env:
  SYSMON_AGENT_PORT   listen port           (default 9888)
  SYSMON_AGENT_BIND   bind address          (default 0.0.0.0 — set to the Tailscale IP to stay on the tailnet)
  SYSMON_AGENT_TOKEN  optional shared secret (?token=.. or Authorization: Bearer ..)

Endpoints:
  GET /metrics (or /)  -> { host, platform, cpu, mem:{pct,usedGB,totalGB}, cores, uptime, ts }
  GET /healthz         -> "ok"
"""
import json
import os
import socket
import threading
import time
from http.server import BaseHTTPRequestHandler, ThreadingHTTPServer
from urllib.parse import urlparse, parse_qs

PORT = int(os.environ.get("SYSMON_AGENT_PORT", "9888"))
BIND = os.environ.get("SYSMON_AGENT_BIND", "0.0.0.0")
TOKEN = os.environ.get("SYSMON_AGENT_TOKEN", "")
GB = 1024 ** 3

_last_cpu = 0.0


def _cpu_times():
    with open("/proc/stat") as f:
        for line in f:
            if line.startswith("cpu "):
                p = [float(x) for x in line.split()[1:]]
                idle = p[3] + (p[4] if len(p) > 4 else 0.0)   # idle + iowait
                return idle, sum(p)
    return 0.0, 0.0


def _sample_loop():
    global _last_cpu
    prev = _cpu_times()
    while True:
        time.sleep(1)
        cur = _cpu_times()
        d_idle, d_total = cur[0] - prev[0], cur[1] - prev[1]
        prev = cur
        if d_total > 0:
            _last_cpu = max(0.0, min(100.0, 100.0 * (1 - d_idle / d_total)))


def _mem():
    info = {}
    with open("/proc/meminfo") as f:
        for line in f:
            k, _, v = line.partition(":")
            try:
                info[k] = float(v.strip().split()[0]) * 1024  # kB -> bytes
            except (IndexError, ValueError):
                pass
    total = info.get("MemTotal", 0.0)
    avail = info.get("MemAvailable", info.get("MemFree", 0.0))
    used = max(0.0, total - avail)
    pct = (used / total * 100) if total else 0.0
    return round(pct, 1), round(used / GB, 2), round(total / GB, 2)


def _uptime():
    try:
        with open("/proc/uptime") as f:
            return int(float(f.read().split()[0]))
    except Exception:
        return 0


def _snapshot():
    pct, used_gb, total_gb = _mem()
    return {
        "host": socket.gethostname(),
        "platform": "linux",
        "cpu": round(_last_cpu, 1),
        "mem": {"pct": pct, "usedGB": used_gb, "totalGB": total_gb},
        "cores": os.cpu_count() or 0,
        "uptime": _uptime(),
        "ts": int(time.time() * 1000),
    }


class Handler(BaseHTTPRequestHandler):
    def log_message(self, *a):
        pass

    def _send(self, code, body, ctype="application/json"):
        b = body.encode() if isinstance(body, str) else body
        self.send_response(code)
        self.send_header("Content-Type", ctype)
        self.send_header("Access-Control-Allow-Origin", "*")
        self.send_header("Content-Length", str(len(b)))
        self.end_headers()
        self.wfile.write(b)

    def _authed(self, q):
        if not TOKEN:
            return True
        if q.get("token", [None])[0] == TOKEN:
            return True
        return self.headers.get("Authorization", "") == "Bearer " + TOKEN

    def do_GET(self):
        u = urlparse(self.path)
        if u.path == "/healthz":
            return self._send(200, "ok", "text/plain")
        if u.path in ("/metrics", "/"):
            if not self._authed(parse_qs(u.query)):
                return self._send(401, json.dumps({"error": "unauthorized"}))
            return self._send(200, json.dumps(_snapshot()))
        self._send(404, json.dumps({"error": "not found"}))


def main():
    threading.Thread(target=_sample_loop, daemon=True).start()
    time.sleep(1.1)   # let the first CPU delta accumulate
    srv = ThreadingHTTPServer((BIND, PORT), Handler)
    print(f"[sysmon-agent.py] listening on {BIND}:{PORT}" + (" (token required)" if TOKEN else ""), flush=True)
    srv.serve_forever()


if __name__ == "__main__":
    main()

File: agent/sysmon-agent.mjs
#!/usr/bin/env node
// sysmon-agent — tiny zero-dependency metrics endpoint for the Ulanzi System
// Monitor plugin's remote (Tailscale) host sources.
//
// Run one of these on each host you want to monitor. It samples CPU + memory
// locally (Node `os`, cross-platform: Linux / macOS / Windows) and serves the
// latest snapshot as JSON over HTTP. The Ulanzi plugin (running on the PC with
// Ulanzi Studio) fetches it over your Tailscale network.
//
//   node sysmon-agent.mjs
//
// Config via env vars:
//   SYSMON_AGENT_PORT   listen port           (default 9888)
//   SYSMON_AGENT_BIND   bind address          (default 0.0.0.0 — restrict via Tailscale ACLs)
//   SYSMON_AGENT_TOKEN  optional shared secret; if set, requests must carry it as
//                       `?token=...` or `Authorization: Bearer ...`
//
// Endpoints:
//   GET /metrics  (alias: GET /)  -> { host, cpu, mem:{pct,usedGB,totalGB}, cores, ts, uptime }
//   GET /healthz                  -> "ok"
//
// No external dependencies — only Node built-ins, so it runs anywhere Node does.

import http from 'http';
import os from 'os';
import fs from 'fs';

// Preferred CPU thermal-zone types, best first. Linux only; null elsewhere.
const TEMP_PRI = ['x86_pkg_temp', 'cpu-thermal', 'cpu_thermal', 'coretemp', 'k10temp', 'soc_thermal', 'soc', 'acpitz'];
function readTemp() {
  if (os.platform() !== 'linux') return null;
  const zones = {};
  let dirs = [];
  try { dirs = fs.readdirSync('/sys/class/thermal'); } catch (e) { dirs = []; }
  for (const d of dirs) {
    if (!d.startsWith('thermal_zone')) continue;
    try {
      const t = fs.readFileSync(`/sys/class/thermal/${d}/type`, 'utf8').trim();
      const v = parseInt(fs.readFileSync(`/sys/class/thermal/${d}/temp`, 'utf8').trim(), 10);
      if (v > 0 && v < 200000 && zones[t] == null) zones[t] = v;
    } catch (e) { /* ignore */ }
  }
  for (const p of TEMP_PRI) if (zones[p] != null) return Math.round(zones[p] / 1000);
  const vals = Object.values(zones);
  return vals.length ? Math.round(Math.max(...vals) / 1000) : null;
}

const PORT = parseInt(process.env.SYSMON_AGENT_PORT || '9888', 10);
const BIND = process.env.SYSMON_AGENT_BIND || '0.0.0.0';
const TOKEN = process.env.SYSMON_AGENT_TOKEN || '';
const GB = 1024 * 1024 * 1024;
const clamp = (v) => Math.max(0, Math.min(100, v));

// --- CPU sampling (idle/total deltas across all cores) -----------------------
function cpuTimes() {
  let idle = 0, total = 0;
  for (const c of os.cpus() || []) {
    for (const k in c.times) total += c.times[k];
    idle += c.times.idle;
  }
  return { idle, total };
}

let prev = cpuTimes();
let lastCpu = 0;
// Sample on a steady internal cadence so CPU% is accurate regardless of how
// often (or rarely) the plugin polls us.
function tick() {
  const t = cpuTimes();
  const dIdle = t.idle - prev.idle;
  const dTotal = t.total - prev.total;
  prev = t;
  if (dTotal > 0) lastCpu = clamp(100 * (1 - dIdle / dTotal));
}
setInterval(tick, 1000).unref();

function snapshot() {
  const total = os.totalmem();
  const free = os.freemem();
  const used = total - free;
  return {
    host: os.hostname(),
    platform: os.platform(),
    cpu: Math.round(lastCpu * 10) / 10,
    mem: {
      pct: Math.round(clamp((used / total) * 100) * 10) / 10,
      usedGB: Math.round((used / GB) * 100) / 100,
      totalGB: Math.round((total / GB) * 100) / 100,
    },
    cores: (os.cpus() || []).length,
    temp: readTemp(),
    uptime: Math.round(os.uptime()),
    ts: Date.now(),
  };
}

// --- auth --------------------------------------------------------------------
function authorized(req, url) {
  if (!TOKEN) return true;
  const q = url.searchParams.get('token');
  if (q && q === TOKEN) return true;
  const h = req.headers['authorization'] || '';
  return h === `Bearer ${TOKEN}`;
}

const server = http.createServer((req, res) => {
  const url = new URL(req.url, 'http://localhost');
  res.setHeader('Access-Control-Allow-Origin', '*');

  if (url.pathname === '/healthz') {
    res.writeHead(200, { 'Content-Type': 'text/plain' });
    return res.end('ok');
  }
  if (url.pathname === '/metrics' || url.pathname === '/') {
    if (!authorized(req, url)) {
      res.writeHead(401, { 'Content-Type': 'application/json' });
      return res.end(JSON.stringify({ error: 'unauthorized' }));
    }
    res.writeHead(200, { 'Content-Type': 'application/json' });
    return res.end(JSON.stringify(snapshot()));
  }
  res.writeHead(404, { 'Content-Type': 'application/json' });
  res.end(JSON.stringify({ error: 'not found' }));
});

server.listen(PORT, BIND, () => {
  // eslint-disable-next-line no-console
  console.log(`[sysmon-agent] listening on ${BIND}:${PORT}` + (TOKEN ? ' (token required)' : ''));
});

process.on('SIGTERM', () => server.close(() => process.exit(0)));
process.on('SIGINT', () => server.close(() => process.exit(0)));
